udaykalra/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -13,11 +13,188 @@
+    /* Create an output collection with a full day to cut(shorten or split)
+     * into the available ranges.
+     */
+    TimeRange fullDay = TimeRange.WHOLE_DAY;
+    Collection<TimeRange> availableRanges = new HashSet<>();
+    availableRanges.add(fullDay);
+
+    // Obtain data from the request.
+    Collection<String> attendees = request.getAttendees();
+    long requestDuration = request.getDuration();
+
+    Collection<TimeRange> busyRanges = new HashSet<>();
+
+    // Detect relevant busy time ranges based on attendees.
+    for (Event event : events) {
+      if (!Collections.disjoint(attendees, event.getAttendees())) {
+        busyRanges.add(event.getWhen());
+      }
+    }
+
+    /* Iteratively cut availableRange's times based on
+     * overlap with busy events and store results in availableRanges.
+     */
+    for (TimeRange busyRange : busyRanges) {
+      // For a single busy event, check all time ranges for overlap and cut
+      removeOverlapsWithTimeRange(busyRange, availableRanges);
+    }
+
+    // Eliminate invalid time intervals by duration.
+    for (TimeRange toEvaluate : availableRanges) {
+      if (toEvaluate.duration() < requestDuration) {
+        availableRanges.remove(toEvaluate);
+      }
+    }
+
+    // Convert to list and sort.
+    List outList = new ArrayList(availableRanges);
+    Collections.sort(outList, TimeRange.ORDER_BY_START);
+    return outList;
+  }
+
+  /**
+   * Mutates timeRanges by removing the time range defined in possiblyOverlappingTimeRange from any
+   * overlapping TimeRange objects it contains.
+   */
+  private void removeOverlapsWithTimeRange(
+      TimeRange possiblyOverlappingTimeRange, Collection timeRanges) {
+
+    // Make a snapshot of TimeRanges for each busy check
+    Collection<TimeRange> rangesSnapshot = new HashSet<>();
+    rangesSnapshot.addAll(timeRanges);
+
+    for (TimeRange currentTimeRange : rangesSnapshot) {
+
+      // Check for overlap and adjust currentTimeRange
+      if ((currentTimeRange.overlaps(possiblyOverlappingTimeRange)) || (possiblyOverlappingTimeRange.overlaps(currentTimeRange))) {
+        timeRanges.remove(currentTimeRange);
+        
+        //Get TimeRange(s) to replace currentTimeRange
+        Collection<TimeRange> replacementRanges = getReplacementForOverlappedTimeRange(possiblyOverlappingTimeRange, currentTimeRange);
+        for(TimeRange replacementRange : replacementRanges){
+          timeRanges.add(replacementRange);
+        }
+      }
+    }
+  }
+
+  /**
+   * Returns a collection of TimeRanges to replace an overlap with overlappingTimeRange. The
+   * replacement omits the overlap through either:
+   *  1. trimming the TimeRange into a shorter one,
+   *  2. splitting the TimeRange and trimming the resulting TimeRanges.
+   * If events do not overlap, no changes will occur.
+   */
+  private Collection<TimeRange> getReplacementForOverlappedTimeRange(
+      TimeRange overlappingTimeRange, TimeRange toReplace) {
```

Comment:
Suggest `toReplace` -> `overlappedTimeRange` for increased readability.

Reply:
Done